darshan-util/pydarshan/darshan/experimental/plots/data_access_by_filesystem.py

```diff
@@ -0,0 +1,524 @@
+
+
+def convert_file_path_to_root_path(file_path: str) -> str:
+    """
+    Parameters
+    ----------
+
+    file_path: a string containing the absolute file path
+
+    Returns
+    -------
+    A string containing the root path.
+
+    Examples:
+    ---------
+
+    >>> filesystem_root = convert_file_path_to_root_path("/scratch1/scratchdirs/glock/testFile.00000046")
+    >>> filesystem_root
+    '/scratch1'
+
+    """
+    path_parts = pathlib.Path(file_path).parts
+    filesystem_root = ''.join(path_parts[:2])
+    return filesystem_root
+
+def convert_file_id_to_path(input_id: float, file_id_dict: Dict[int, str]) -> Optional[str]:
+    """
+    Parameters
+    ----------
+
+    input_id: a float representing the file hash
+
+    file_id_dict: a dictionary mapping integer file hash values
+                  to string values corresponding to their respective
+                  paths
+
+    Returns
+    -------
+    A string containing the file path path corresponding to ``input_id``,
+    or ``None`` if no matching file path was found for the input hash.
+
+    Examples
+    --------
+
+    >>> # file_id_dict typically comes from `report.data["name_records"]`
+    >>> file_id_dict = {210703578647777632: '/yellow/usr/projects/eap/users/treddy/simple_dxt_mpi_io_darshan/test.out.locktest.0',
+    ...                 9457796068806373448: '/yellow/usr/projects/eap/users/treddy/simple_dxt_mpi_io_darshan/test.out'}
+    >>> # input_id may came from i.e., a pandas dataframe of the record data
+    >>> result = convert_file_id_to_path(input_id=9.457796068806373e+18, file_id_dict=file_id_dict)
+    >>> result
+    '/yellow/usr/projects/eap/users/treddy/simple_dxt_mpi_io_darshan/test.out'
+
+    """
+    for file_id_hash, file_path in file_id_dict.items():
+        if np.allclose(input_id, file_id_hash):
+            return file_path
+    return None
+
+def identify_filesystems(file_id_dict: Dict[int, str], verbose: bool = False) -> List[str]:
+    """
+    Parameters
+    ----------
+    file_id_dict: a dictionary mapping integer file hash values
+                  to string values corresponding to their respective
+                  paths
+
+    verbose: if ``True``, will print the filesystem root paths that
+             are identified
+
+    Returns
+    -------
+    A list of strings containing the unique filesystem root paths
+    parsed from the input dictionary.
+
+    Examples
+    --------
+
+    >>> # file_id_dict is typically from report.data["name_records"]
+    >>> file_id_dict = {210703578647777632: '/yellow/usr/projects/eap/users/treddy/simple_dxt_mpi_io_darshan/test.out.locktest.0',
+    ...                 14388265063268455899: '/tmp/ompi.sn176.28751/jf.29186/1/test.out_cid-0-3400.sm'}
+
+    >>> filesystem_roots = identify_filesystems(file_id_dict=file_id_dict, verbose=True)
+    filesystem_roots: ['/yellow', '/tmp']
+    """
+    filesystem_roots = []
+    excluded = ['<STDIN>', '<STDOUT>', '<STDERR>']
```

Comment:
nobody has commented on this, but I'm curious if we actually want to ignore stdout/stderr and so on?

Since Phil changed the plot section to "category" instead of "filesystem" proper, and because a user may want to know if they're writing 500 GB of data to stderr I suppose?

@carns @nawtrey

Reply:
I guess keeping stderr/stdout/stdin as a row in this table would make some sense? We kind of indirectly are doing the same thing in the current darshan-job-summary tool, except those streams are accounted for as "UNKOWN" in the file system table (that's just what the current code does since there is no associated "mount point" for this data). We could make them a more explicit row in the table since some apps really do do a lot of I/O to standard streams? I'm not sure exactly how to label but it does seem logical to include them.

Reply:
I've pushed in a revision/tests to try to address this. Each `STD..` stream is labelled as its own row in the table of plots. If the `STD..` streams are anonymized, then they are labelled as such:

![test1 png_data_access_by_category](https://user-images.githubusercontent.com/7903078/129258040-061593d3-4f2c-4b3e-b067-e49a091c5f5b.png)
![test2 png_data_access_by_category](https://user-images.githubusercontent.com/7903078/129258043-41f422aa-a596-4de3-bcc3-08c1ce509012.png)

Reply:
Generating a sample log file for the logs repo that has (known) bytes read/written to the std streams seems like a useful thing for someone to do (or open an issue for at least). Perhaps such a file is already in that repo.

Indeed, since these streams are not files in the conventional sense, would be good to test with such a log file to ensure that the control flow doesn't ignore "non-file" scenarios for reporting purposes.

Reply:
(it may also only show up in `STDIO` module if there is activity on those streams? not sure--for someone to look into I I suppose, doesn't seem too hard to check..)

Reply:
Really going back here, but in the anonymized case you share above, how do you know what the standard stream records are to re-label them as "anonymized"? The record name has been turned into a hash, so not sure how you could work back to which records represent STDERR, etc.

Reply:
I can help dig up a log file for this test case if it's still helpful.

And, for the record, it's possible for some of the standard stream records to be accessed via the POSIX module, too, in very rare cases. I think I can dig up a recent example I saw of this, but would be good to ensure we account for it.